Add basic sharness tests

chriscool opened the pull request:
This is to start testing js-ipfs in the same way as go-ipfs.
See issue #412 (Sharness tests).

diasdavid commented:
\o/ Thank you @chriscool :D 

Do all of these run for you? Found any issues? Can I get a status of this endeavour?

I know you have been trying to reach me through IRC, unfortunately, with DEVCON2 I wasn't the most available. I'll have good wifi when the 29th of September comes.

chriscool commented:
There are a few that don't pass. For example I think the first to fail is the 5st or the 6st because `jsipfs version --all` doesn't work properly.
If you want I can change the last commit to use test_expect_failure for those that don't pass.
But I think it can still be interesting to commit this as it makes it possible for other developers and CI tools to try to run the tests and see what fails.

The Makefile also implements `make help` that provides a bit of help to run the tests.

dignifiedquire commented:
I think it would be good to leave the failing as failing for now so we can make a list and check things off :)

diasdavid commented:
Yeah, add all the tests @chriscool, if they fail, the better, more stuff for us to fix :D 

@chriscool idealy, we want:

- A PR that sets up the sharding tests (the ones that pass already) and then
- PR for each of the subcommands, with all the tests, highlighting the ones that fail

chriscool commented:
@diasdavid ok, I removed failing tests from t0010, so now everything should pass and you can merge this.
I am preparing other PRs, some that add passing tests, some that add failing tests.
I will send them when this PR will be merged, as they are based on it.

chriscool commented:
By the way, tests should either be run using `make` or `make deps` should be used before running the tests, and the path to 'js-ipfs/test/bin' should be in your PATH.

This is because a symbolic link from 'js-ipfs/test/bin/ipfs' to 'js-ipfs/src/cli/bin.js' is used to make it possible for the tests to run without having to change them much.

chriscool commented:
This last commit adds a check and improves the error message.

diasdavid commented on test/ipfs-test-lib.sh:
```diff
@@ -0,0 +1,107 @@
+# Generic test functions for go-ipfs
```
I guess we can start calling it the IPFS cli :)

chriscool replied:
Year, sure I will change that.

chriscool replied:
Changed to: `# Generic test functions for ipfs cli tests`

diasdavid commented on test/sharness/Makefile:
```diff
@@ -0,0 +1,53 @@
+	@which curl >/dev/null || (echo "Please install curl!" && false)
+
+.PHONY: all help clean clean-test-results $(T) aggregate deps sharness
+
```
can we bring the Makefile to the root, so we can run it on CI by adding a line with `make` to:

https://github.com/ipfs/js-ipfs/blob/master/.travis.yml#L14

chriscool replied:
In go-ipfs there is also a Makefile at the root, but it is different.
You use `make test` at the root and `make` when you are in test/sharness.
I can do something like that but it could go into another PR I think as it is related to the whole project.

diasdavid replied:
@chriscool what I'm looking for is to be able to add these sharness tests to CI before this merge.

chriscool replied:
Yes, with the last commit I added you can just run `make test` at the root.
I will add that to the '.travis.yml'.

chriscool replied:
I added that and it works.

diasdavid commented on test/sharness/lib/test-lib-hashes.sh:
```diff
@@ -0,0 +1,7 @@
+HASH_WELCOME_DOCS="QmYwAPJzv5CZsnA625s3Xf2nemtYgPpHdWEz79ojWnPbdG"
+HASH_GATEWAY_ASSETS="QmXB7PLRWH6bCiwrGh2MrBBjNkLv3mY3JdYXCikYZSwLED"
+HASH_HELP_PAGE="QmY5heUM5qgRubMDD1og9fhCPA6QdkMp3QCwd4s7gJsyE7"
+HASH_EMPTY_DIR="QmUNLLsPACCz1vLxQVkXqqLX5R1X345qqfHbsf67hvA3Nn"
```
Apparently, these change often (//cc @lgierth) how does go-ipfs handle it? Changing it manually everytime?

chriscool replied:
It doesn't change so often, so I think manually is fine for now.
Last time they changed was in January.

whyrusleeping replied:
We update them manually here every time we update them in go-ipfs

diasdavid commented:
\o/ woot! 

There is a lot of bash magic, but they do work for me, so awesome! :D 

- [ ] small request https://github.com/ipfs/js-ipfs/pull/504/files#r81572723

Once we have CI running these, I'm happy to merge. Then we can see all the remaining PR's for each subcommand coming in :)

Super excited about this, getting the same level of robustness as the Go implementation 🌟

chriscool commented:
@diasdavid just pushed with the following changes:

- added a Makefile at the root as in the last commit,
- changed the first line of test/ipfs-test-lib.sh

chriscool commented:
@diasdavid could you take a look at merging this?

diasdavid commented:
@chriscool done, thank you :) 

Can you start opening the PR for each of the subcommands now? If possible, listing which tests fail, so that we can fix them :)

chriscool commented:
@diasdavid ok I will do that now.
